Fix lag when scrolling horizontally to seek, #4153

low-batt opened the pull request:
This commit will:
- Add methods displayActive and displayIdle to VideoView
- Add use of a Timer to delay stopping the display link
- Change MPVController to ensure display link is running while seeking
- Change PlayerCore to ensure display link is running while stepping
- Change MainWindowController to ensure display link is running when entering and leaving full screen mode

- [x] I have read [CONTRIBUTING.md](https://github.com/iina/iina/blob/develop/CONTRIBUTING.md)
- [x] This implements/fixes issue #4153.

---

**Description:**

low-batt committed:
Fix lag when scrolling horizontally to seek, #4153

This commit will:
- Add methods displayActive and displayIdle to VideoView
- Add use of a Timer to delay stopping the display link
- Change MPVController to ensure display link is running while seeking
- Change PlayerCore to ensure display link is running while stepping
- Change MainWindowController to ensure display link is running when
  entering and leaving full screen mode

Rebased with develop and corrected merge conflicts.

low-batt commented:
See the issue for an analysis of the problem.

There one case the proposed fix does not cover. In order to properly support the user sending commands directly to mpv through the IPC interface, control of the display link should be based on events from mpv. This works for seeking:
```bash
echo '{ "command": ["seek", "5"] }' | socat - /tmp/mpv-socket | jq
```

It does not work for stepping:
```bash
echo '{ "command": ["frame-step"] }' | socat - /tmp/mpv-socket | jq
```

Directly commanding mpv to step through the IPC interface works, but it does cause mpv to report "mpv_render_report_swap() not being called". This is because the code that insures the display link is working for stepping is in `PlayerCore.frameStep` and not in `MPVController` in response to a mpv event. That is because I was unable to find a mpv event that IINA can key off of for stepping.

low-batt commented:
Rebased with develop and corrected merge conflicts.

uiryuu commented on iina/PlayerCore.swift:
```diff
@@ -668,11 +668,18 @@ class PlayerCore: NSObject {
     } else {
       mpv.command(.frameStep)
     }
+    if info.isPaused {
```
After you send `.frameStep` or `.frameBackStep` to mpv, the playback will always stop, regardless the playback status before the command.

uiryuu commented on iina/PlayerCore.swift:
```diff
@@ -668,11 +668,18 @@ class PlayerCore: NSObject {
+    // When playback is paused the display link is stopped in order to avoid wasting energy on
+    // It must be running when stepping to avoid slowdowns caused by mpv waiting for IINA to call
+    // mpv_render_report_swap.
+    mainWindow.videoView.displayActive()
```
It seems even if the display link is not running, frame step commands can run correctly. Do we have to activate the display link manually?

low-batt replied:
Stepping, seeking, all currently work because mpv will timeout and continue on. When that happens mpv will complain:
```text
[  14.169][v][vo/libmpv] mpv_render_report_swap() not being called.
```

We can tolerate some of this happening, but IINA should always try and avoid it. We don't want to see lots of these warnings from mpv when IINA is working correctly. For one thing that could hide other problems that trigger this warning that should not be occurring.

Because mpv does not give us an event to key off of (at least not that I could find) we have to tolerate this situation if the user sends a step command behind the back of IINA directly to mpv using the IPC interface. Unfortunate, but I don't see a way to deal with that situation.

low-batt committed:
Fix lag when scrolling horizontally to seek, #4153

This commit will:
- Add methods displayActive and displayIdle to VideoView
- Add use of a Timer to delay stopping the display link
- Change MPVController to ensure display link is running while seeking
- Change PlayerCore to ensure display link is running while stepping
- Change MainWindowController to ensure display link is running when
  entering and leaving full screen mode

Rebased with develop and corrected merge conflicts.
Corrected PlayerCore.frameStep as per review comment.

svobs commented:
Very good fix! Looks like it has secondary improvements too like snappier & more reliable OSD updating.
